RIOT-OS/RIOT
# Makefile.features: add declarative FEATURES_ variables definition

### Contribution description

Use `Makefile.features` to define all the calculated `FEATURES_` variables.
This gives a name on the resolved parsing of the BSP `FEATURES_` variables.

Then use them in `Makefile.include`/`makefiles/info.inc.mk`/makefiles/info-global.inc.mk`.

### Reviewing procedure

The review is mainly on `Makefile.features` the other changes being the consequence of the changes but they can explain what changed.

The goal is does the name makes sense and matches with what is evaluated.

The fact that the migration is correct can come in a second time.

### Testing procedure

All the supported applications should still be the same.

I added `make info-applications-supported-boards` that can be run in the base directory to compare a commit with another for listing all the boards returned by `info-boards-supported`. (I will split this one).

The output of this one should be the same with the end commit and the commit introducing it.

#### Info-build output

I will add some outputs for the `info-build` to verify the new variables in some defined cases.

Difference between master and this pull request

```diff
diff -u  info_build_examples_default_samr21_xpro_*
--- info_build_examples_default_samr21_xpro_master      2019-05-27 17:47:39.169868988 +0200
+++ info_build_examples_default_samr21_xpro_pr  2019-05-27 17:47:18.273697122 +0200
@@ -21,19 +21,23 @@
 HEXFILE: /home/harter/work/git/RIOT/examples/default/bin/samr21-xpro/default.hex
 FLASHFILE: /home/harter/work/git/RIOT/examples/default/bin/samr21-xpro/default.bin
 
-FEATURES_REQUIRED (excl. optional features):
+FEATURES_USED:
+         periph_cpuid periph_gpio periph_gpio_irq periph_pm periph_rtc periph_spi periph_timer periph_uart
+FEATURES_REQUIRED:
          periph_gpio periph_gpio_irq periph_spi periph_timer periph_uart
-FEATURES_OPTIONAL (strictly "nice to have"):
+FEATURES_OPTIONAL_ONLY (optional that are not required, strictly "nice to have"):
          periph_cpuid periph_hwrng periph_pm periph_rtc
+FEATURES_OPTIONAL_MISSING (missing optional features):
+         periph_hwrng
 FEATURES_PROVIDED (by the board or USEMODULE'd drivers):
          cpp cpu_check_address periph_adc periph_cpuid periph_flashpage periph_flashpage_raw periph_flashpage_rwee periph_gpio periph_gpio_irq periph_i2c periph_pm periph_pwm periph_rtc periph_rtt periph_spi periph_timer periph_uart periph_usbdev puf_sram riotboot
-FEATURES_MISSING (incl. optional features):
-         -none-
-FEATURES_MISSING (only non-optional features):
+FEATURES_MISSING (only non optional features):
          -none-
 
 FEATURES_CONFLICT:     
 FEATURES_CONFLICT_MSG: 
+FEATURES_CONFLICTING:
+         -none-
 
 INCLUDES: 
        -isystem  
```


```
BOARD=samr21-xpro make info-build | grep -C 1 'FEATURES' # master

FEATURES_REQUIRED (excl. optional features):
         periph_gpio periph_gpio_irq periph_spi periph_timer periph_uart
FEATURES_OPTIONAL (strictly "nice to have"):
         periph_cpuid periph_hwrng periph_pm periph_rtc
FEATURES_PROVIDED (by the board or USEMODULE'd drivers):
         cpp cpu_check_address periph_adc periph_cpuid periph_flashpage periph_flashpage_raw periph_flashpage_rwee periph_gpio periph_gpio_irq periph_i2c periph_pm periph_pwm periph_rtc periph_rtt periph_spi periph_timer periph_uart periph_usbdev puf_sram riotboot
FEATURES_MISSING (incl. optional features):
         -none-
FEATURES_MISSING (only non-optional features):
         -none-

FEATURES_CONFLICT:     
FEATURES_CONFLICT_MSG: 

```

```
BOARD=samr21-xpro make info-build | grep -C 1 'FEATURES' # PR

FEATURES_USED:
         periph_cpuid periph_gpio periph_gpio_irq periph_pm periph_rtc periph_spi periph_timer periph_uart
FEATURES_REQUIRED:
         periph_gpio periph_gpio_irq periph_spi periph_timer periph_uart
FEATURES_OPTIONAL_ONLY (optional that are not required, strictly "nice to have"):
         periph_cpuid periph_hwrng periph_pm periph_rtc
FEATURES_OPTIONAL_MISSING (missing optional features):
         periph_hwrng
FEATURES_PROVIDED (by the board or USEMODULE'd drivers):
         cpp cpu_check_address periph_adc periph_cpuid periph_flashpage periph_flashpage_raw periph_flashpage_rwee periph_gpio periph_gpio_irq periph_i2c periph_pm periph_pwm periph_rtc periph_rtt periph_spi periph_timer periph_uart periph_usbdev puf_sram riotboot
FEATURES_MISSING (only non optional features):
         -none-

FEATURES_CONFLICT:     
FEATURES_CONFLICT_MSG: 
FEATURES_CONFLICTING:
         -none-
```

Output with conflicting features:

```
BOARD=stm32f4discovery make -C tests/warn_conflict/ info-build | grep -C 1 'FEATURES' 

FEATURES_USED:
         periph_dac periph_gpio periph_pm periph_spi periph_uart
FEATURES_REQUIRED:
         periph_dac periph_spi periph_uart
FEATURES_OPTIONAL_ONLY (optional that are not required, strictly "nice to have"):
         periph_gpio periph_pm
FEATURES_OPTIONAL_MISSING (missing optional features):
         -none-
FEATURES_PROVIDED (by the board or USEMODULE'd drivers):
         arduino cpp cpu_check_address periph_adc periph_cpuid periph_dac periph_gpio periph_gpio_irq periph_hwrng periph_i2c periph_pm periph_pwm periph_rtc periph_spi periph_timer periph_uart periph_uart_modecfg puf_sram
FEATURES_MISSING (only non optional features):
         -none-

FEATURES_CONFLICT:     periph_spi:periph_dac
FEATURES_CONFLICT_MSG: "On stm32f4discovery boards there are the same pins for the DAC and/or SPI_0."
FEATURES_CONFLICTING:
         periph_dac periph_spi
```

### Issues/PRs references

Side work of https://github.com/RIOT-OS/RIOT/issues/9913 that added this `Makefile.features` to group the `features` handling.

Could help with https://github.com/RIOT-OS/RIOT/pull/9081 to have the definition in only one place.

* [x] `info-applications-supported-boards` https://github.com/RIOT-OS/RIOT/pull/11590
